### PR TITLE
Authorize active_admin_comments in resource#show

### PR DESCRIPTION
**Pull request by amiuhle**
Check whether user is authorized to view comments when displaying them in resources#show.

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
@amiuhle We no longer maintain 1-3-stable, is this issue present in the master branch?

**Comment by deivid-rodriguez** [comment]
Thank you! Could you add a test for this?

**Comment by amiuhle** [comment]
Added two feature specs.

Would you consider cherry-picking and releasing this as `1.3.2`?

File: lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb [hunk 1]
```diff
@@ -11,10 +11,12 @@ class Comments < ActiveAdmin::Views::Panel
         attr_accessor :resource
 
         def build(resource)
-          @resource = resource
-          @comments = ActiveAdmin::Comment.find_for_resource_in_namespace(resource, active_admin_namespace.name).includes(:author).page(params[:page])
-          super(title, for: resource)
-          build_comments
+          if authorized?(ActiveAdmin::Auth::READ, ActiveAdmin::Comment)
+            @resource = resource
+            @comments = active_admin_authorization.scope_collection(ActiveAdmin::Comment.find_for_resource_in_namespace(resource, active_admin_namespace.name).includes(:author).page(params[:page]))
```
Comment (Fivell) [inline lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb]: @amiuhle there is also ```build_comment_form``` maybe u need to wrap it with
```
if authorized?(ActiveAdmin::Auth::CREATE, ActiveAdmin::Comment)

end
```

File: features/comments/commenting.feature [hunk 2]
```diff
@@ -5,7 +5,11 @@ Feature: Commenting
   I want to store and view comments on a resource
 
   Background:
-    Given a post with the title "Hello World" written by "Jane Doe" exists
+    Given a configuration of:
+    """
+      ActiveAdmin.application.namespace(:admin).authorization_adapter = ActiveAdmin::AuthorizationAdapter
+    """
+    And a post with the title "Hello World" written by "Jane Doe" exists
```
Comment (deivid-rodriguez) [inline features/comments/commenting.feature]: You can leave this setup untouched and instead tag each of the new specs with the [`@authorization` tag](https://github.com/activeadmin/activeadmin/blob/9292dfb548b7779db962ab7992ea4b1342ea69d4/features/support/env.rb#L110-L115).

File: features/comments/commenting.feature [hunk 3]
```diff
@@ -176,3 +180,74 @@ Feature: Commenting
     And I should see 20 comments
     And I should see "Displaying comments 51 - 70 of 70 in total"
     And I should not see the pagination "Next" link
+
+  Scenario: Not authorized to list comments
+    Given 5 comments added by admin with an email "commenter@example.com"
+    And 3 comments added by admin with an email "admin@example.com"
+    And a show configuration of:
+      """
+        class ListCommentsForCommenterOnly < ActiveAdmin::AuthorizationAdapter
```
Comment (deivid-rodriguez) [inline features/comments/commenting.feature]: This class name does not match what it does, maybe better `NeverListComments` or something like that?

File: features/comments/commenting.feature [hunk 4]
```diff
@@ -176,3 +180,74 @@ Feature: Commenting
...
+
+        ActiveAdmin.application.namespace(:admin).authorization_adapter = CommentsByCommenterOnly
+
+        ActiveAdmin.register Post
+      """
+    Then I should see "Comments (3)"
+    And I should see 3 comments
+    And I should be able to add a comment
+
+  Scenario: Authorized to list and view own comments
```
Comment (deivid-rodriguez) [inline features/comments/commenting.feature]: This scenario description does not match what the scenario does. Maybe better "Not authorized to create comments"?

File: features/comments/commenting.feature [hunk 5]
```diff
@@ -176,3 +180,74 @@ Feature: Commenting
...
+        end
+
+        ActiveAdmin.application.namespace(:admin).authorization_adapter = NoNewComments
+
+        ActiveAdmin.register Post
+      """
+    Then I should see "Comments (5)"
+    And I should see 5 comments
+    And I should not be able to add a comment
+
```
Comment (deivid-rodriguez) [inline features/comments/commenting.feature]: Can you remove the trailing blank line?

**Comment by amiuhle** [comment]
@deivid-rodriguez I had changed the functionality of `ListCommentsForCommenterOnly` during development, instead of renaming it, I changed it again to match the class name.

Everything else is resolved, and I rebased a bit.

**Comment by amiuhle** [comment]
@deivid-rodriguez Using `collection.where(author: user)` now instead of finding the user of a specific email address. That was kind of stupid anyways :D

I did another rebase on master, so the changes you made in #5585 are gone.

Let me know whether you'd like me to rebase the changes from the last commit into the corresponding previous commits.

**Comment by deivid-rodriguez** [comment]
If I was to cleanup the history of this PR, I would probably squash everything into one commit, up to you!

**Comment by amiuhle** [comment]
@deivid-rodriguez Should I add documentation? I think I could provide a `PunditAdapter` example.

Also, could you take a look at #5554?

**Comment by deivid-rodriguez** [comment]
Sure, yes!

**Comment by deivid-rodriguez** [comment]
> I'll squash them. Just in case you didn't know this, you can do that via GitHub if you enable it in the repository settings: https://blog.github.com/2016-04-01-squash-your-commits/

Yeah, I think we agreed a while ago to not enable this merge option and consistently use the standard merge, but I would personally enable it: it's handy and it helps in cases like this one.

File: lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb]: I had the same thought and... I'm not sure, I think scoping the collection is ok.

File: lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb [hunk 6]
```diff
@@ -11,10 +11,12 @@ class Comments < ActiveAdmin::Views::Panel
         attr_accessor :resource
 
         def build(resource)
-          @resource = resource
-          @comments = ActiveAdmin::Comment.find_for_resource_in_namespace(resource, active_admin_namespace.name).includes(:author).page(params[:page])
-          super(title, for: resource)
-          build_comments
+          if authorized?(ActiveAdmin::Auth::READ, ActiveAdmin::Comment)
```
[the remark this answers is not in the document]
Comment (deivid-rodriguez) [reply lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb]: That's interesting, yeah, open an issue :+1:.

**Comment by amiuhle** [comment]
@deivid-rodriguez I expanded on the existing pundit templates from the specs, since they're already referenced in the doumentation.

I'm not 100% happy with what the `ActiveAdmin::CommentPolicy` does vs how it's tested, but since the logged in user can't view any comments from other users, I didn't know how to test the `delete?` policy, for example.

I don't want to spend much more time, if you think it's okay like this, or have a good idea how to improve the last commit, I'll add some modifications. Otherwise I'll just remove the last commit and squash so you can merge this.

**Pull request merged**